src/main/scala/srdc/smartcds/cds/flow/ACCAHAFlowExecution.scala

```diff
@@ -0,0 +1,395 @@
+    val ethnicityCodes = ethnicity.flatMap(_.valueCodeableConcept.toSeq).flatMap(_.coding.map(_.code))
+
+    /* Paper only cares whether the patient is black or not, so did I */
+    if (blackEthnicityCodes.intersect(ethnicityCodes).nonEmpty) "africanamerican" else "white"
+  }
+
+  /**
+   * Determines the smoking status of the patient
+   *
+   * @param smokingObs patient's smoking observation
+   * @return patient's current smoking status, 1 if smokes and 0 otherwise
+   */
+  //noinspection DuplicatedCode
+  private def determineSmokingStatus(smokingObs: Option[Observation]): Int = {
+    val smoking = if (smokingObs.isDefined && smokingObs.get.valueCodeableConcept.isDefined) {
+      smokingObs.get.valueCodeableConcept.get.coding.map(_.code).toSeq
+    } else {
+      Seq("266919005")
+    }
+
+    val smoke_cat = if (Seq("LA18978-9", "LA18980-5", "266919005").intersect(smoking).nonEmpty) 0
+    else if (smoking.contains("LA15920-4", "8517006")) 1
+    else if (Seq("LA18977-1", "LA18982-1").intersect(smoking).nonEmpty) 2
+    else if (Seq("LA18979-7", "LA18976-3", "449868002").intersect(smoking).nonEmpty) 3
+    else if (smoking.contains("LA18981-3")) 4
+    else 0
+
+    /* Treat "never smoked" and "former smoker" as 0, others as 1 */
+    if (smoke_cat == 0 || smoke_cat == 1) 0 else 1
+  }
+
+  /**
+   * Validates given prefetch and returns the ACC/AHA risk score
+   *
+   * @param patient               Patient resource
+   * @param TotalCholesterol      Total Cholesterol Observation
+   * @param HDLCholesterol        HDL Observation
+   * @param SystolicBP            Blood Pressure Observation
+   * @param SmokingStatus         Smoking Status Observation
+   * @param Type1Diabetes         Type 1 Diabetes Condition
+   * @param Type2Diabetes         Type 2 Diabetes Condition
+   * @param HypertensiveTreatment Hypertensive Treatment Medication Statement
+   * @param Ethnicity             Ethnicity Observation for Patient
+   * @return A double tuple consisting of patient's risk score and healthy score of a hypothetical
+   *         patient with same age, sex, gender, but with optimal health parameters
+   */
+  private def calculateACCRisk(patient: Patient, TotalCholesterol: Seq[Observation], HDLCholesterol: Seq[Observation],
+                               SystolicBP: Seq[Observation], SmokingStatus: Seq[Observation], Type1Diabetes: Seq[Condition], Type2Diabetes: Seq[Condition],
+                               HypertensiveTreatment: Seq[MedicationStatement], Ethnicity: Seq[Observation]): Option[(Double, Double)] = {
+
+    /* Get related information about the patient */
+    val age = FhirParseHelper.getAge(patient)
+    val gender = patient.gender
+    val raceOpt = Option(determineRace(Ethnicity))
+    val raceObs = Ethnicity.headOption
+    val diabetes = checkExists(Type1Diabetes) | checkExists(Type2Diabetes)
+    val treatedHypertension = checkHypertensiveTreatment(HypertensiveTreatment)
```

Comment:
Why did you not use `checkExists(HypertensiveTreatment)` instead of `checkHypertensiveTreatment(HypertensiveTreatment)`? It's already defined in the prefetch with related hypertensive treatment codes, do we need to check the codes again?

Reply:
Fixed.